packages/forms/src/UIForm/UIForm.component.js

```diff
@@ -46,11 +44,31 @@ export default class UIForm extends React.Component {
 	onChange(event, { schema, value }) {
 		const payload = {
 			formName: this.props.formName,
-			properties: this.props.properties,
+			properties: this.props.properties || {},
 			schema,
 			value,
 		};
-		this.props.onChange(event, payload);
+		payload.formData = mutateValue(payload.properties, schema.key, value);
+		if (this.props.moz) {
+			this.props.onChange(payload);
+		} else {
+			this.props.onChange(event, payload);
+		}
+		// trigger if value is correct
```

Comment:
moving trigger in `onChange` implies that trigger will be triggered even if value is not valid
onChange/onFinish separation was to allow widgets and fields to manage when those are called. Some widgets call onChange then onFinish on the same event.
We can for example on selection widget, to trigger both, but on user typing widgets, call onFinish on blur

Reply:
I don't think we have this use case.
All our triggers are on controlled field (like select). So result can't be invalid and happens on change.
I guess trigger the validation onChange will slow down the rendering which is already quite slow.